server/core_storage.go

```diff
@@ -698,11 +698,15 @@ func storagePrepBatch(batch *pgx.Batch, authoritativeWrite bool, op *StorageOpWr
 		// Existing permission checks are not applicable for new storage objects.
 		query = `
 		INSERT INTO storage (collection, key, user_id, value, version, read, write, create_time, update_time)
-		VALUES ($1, $2, $3, $4, $5, $6, $7, now(), now())
+		SELECT $1, $2, $3, $4, $5, $6, $7, now(), now()
+			WHERE NOT EXISTS(SELECT 1 FROM storage WHERE collection = $1 AND key = $2 AND user_id = $3) -- avoids ON CONFLICT hitting pre-existing object, and immediately cancels insert
+		ON CONFLICT (collection, key, user_id) DO UPDATE
```

Comment:
`ON CONFLICT ` misses `WHERE` to check that conflictings row are in fact the same, otherwise this query just behaves as "last  concurrent write wins"

Reply:
`WHERE` clause goes after `ON CONFLICT DO UPDATE SET`